torchtext/prototype/generate.py

```diff
@@ -0,0 +1,324 @@
+            if self.is_huggingface_model:
+                model_inputs["return_dict"] = True
+                model_inputs["output_hidden_states"] = True
+
+            # Get model output
+            outputs = self.model(**model_inputs)
+            output_key = "logits" if self.is_huggingface_model else "decoder_output"
+            decoder_output = outputs[output_key]
+
+            # Calculate probabilities and take the most likely next token
+            # Why do we take the last token instead of a mean_pooling across all of them?
+            probs = F.log_softmax(decoder_output[:, -1], dim=-1)
+            _, next_tokens = torch.topk(probs, 1)
+
+            # For any finished sequences, padding idx should be the last token
+            if eos_idx is not None:
+                if pad_idx is not None:
+                    next_tokens = next_tokens * unfinished_sequences + pad_idx * (1 - unfinished_sequences)
+
+            # Append the next tokens to the previous tokens
+            input_ids = torch.cat([input_ids, next_tokens], dim=-1)
+
+            if eos_idx is not None:
+                unfinished_sequences = unfinished_sequences.mul((next_tokens != eos_idx).long())
+
+            # Stop iterating once all sequences are finished or exceed the max_len
+            if unfinished_sequences.max() == 0 or len(input_ids[0]) >= max_len:
+                break
+
+        return input_ids
+
+    def beam_search(
+        self, 
+        input_ids: torch.Tensor,
+        num_beams: int,
+        max_len: int,
+        beam_size_token: int,
+        beam_threshold: int,
+        eos_score: float,
+        eos_idx: int,
+        num_python_workers: int,
+        **model_kwargs
+    ) -> torch.Tensor:
+        """Beam search implemented using Flashlight Text (https://github.com/flashlight/text).
+
+        Args:
+            input_ids (Tensor): Tokenized startings sequence(s).
+            num_beams (int): Number of beams to use in the beam search.
+            max_len (int): Maximum number of tokens to generate.
+            beam_size_token (int): Vocab size for the LM being used.
+            beam_threshold (int): Threshold before pruning.
+            eos_score (float): Score to input when `eos_idx` is generated.
+            eos_idx (int): End-of-sequence index.
+            num_python_workers (int): Number of python workers to use for multiprocessing.
+            **model_kwargs
+        
+        Returns:
+            Tensor of the generated sequences.
+        """
+        if self.is_encoder_decoder:
+            encoder_output_key = "last_hidden_state" if self.is_huggingface_model else "encoder_output"
+            encoder_output = model_kwargs["encoder_outputs"][encoder_output_key]
+
+        def update_func(emissions, N, T, prev_step_token_idxs, prev_step_model_states, timestep):
+            # `emissions` and `N` are unused in this current implementation
```

Comment:
Could change this around to take in emissions, but AFAIK there is no easy way to get the actual tensor from the data_ptr.